framework-12th-gen-intel: Refactor

SebTM opened the pull request:
###### Description of changes
Kernel: 6.8.2
Tested with: https://github.com/NixOS/nixpkgs/pull/300781

Came across this due to backlight not restored after reboot (https://github.com/NixOS/nixpkgs/issues/178345 - wasn't related) issue and followed as I have the same issue as (https://github.com/NixOS/nixpkgs/issues/202846#issue-1464490034) 2. with the screen "flickering" (like restarting once) and then weird display issue instead like above.

With this changes and above configuration it looks like t his on initrd luks-unlock (how it should be):
![IMG_0863](https://github.com/NixOS/nixos-hardware/assets/17243347/88d5cfd3-7048-4d88-bb4a-18cf1702aecc)

Previous picture:
![photo_2024-04-08_10-09-16](https://github.com/NixOS/nixos-hardware/assets/17243347/75f59f4e-d149-4561-baed-bcdf875d79c2)
(Display issues on current master with i915 loaded in initrd)

Superseeds https://github.com/NixOS/nixos-hardware/pull/846 (fixed and package updated https://github.com/NixOS/nixpkgs/pull/300781)

Fixes https://github.com/NixOS/nixos-hardware/pull/897 for latest kernel-version (6.8) - might need to be removed completely (https://github.com/NixOS/nixos-hardware/issues/672#issuecomment-2015778670) ℹ️ (needs feedback)

Fixes https://github.com/NixOS/nixos-hardware/issues/894

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested the changes in your own NixOS Configuration
- [x] Tested the changes end-to-end by using your fork of `nixos-hardware` and
      importing it via `<nixos-hardware>` or Flake input

RaitoBezarius commented:
@mergify rebase

RaitoBezarius commented:
@mergify rebase

RaitoBezarius commented on common/gpu/intel/default.nix:
```diff
@@ -1,15 +1,26 @@
-
-  environment.variables = {
-    VDPAU_DRIVER = lib.mkIf config.hardware.opengl.enable (lib.mkDefault "va_gl");
+  options.hardware.intelgpu.loadInInitrd = lib.mkEnableOption (lib.mdDoc
```
Is there a value of making this an option? People will probably not discover it.

SebTM replied:
Saw it was done like this in the amd-module and encouraged in some issues to use options - I don't mind just removing it from framework but maybe not everyone is happy - this way it keeps backwards compatible...

RaitoBezarius replied:
OK

RaitoBezarius commented:
Could you clean up the commits and split them in functional logical units? It's hard to review this refactor otherwise.

RaitoBezarius commented:
It's fine!

SebTM commented:
Hey, I've spitted it to the best of abilities but the refactoring in the 12th-gen-intel is hard to split - is that fine otherwise I have to do another round on it? I referenced all issues it fixes (which should explain the changes), added comments for my test-results for the options in questions.

Also I tested all FN-Keys (like sound, brightness, airplane mode) they still work for me. As I'm on 6.8 kernel and not have the time to test lower ones I have set the boundary for my changes to 6.8 - if someone want's/tests it on lower version I'm happy to adjust.

huantianad commented:
I was previously trying to figure this out in the oneVPL thread, but should we perhaps also add onevpl-intel-driver, and replace intel-media-driver with it? Or are they both needed?

RaitoBezarius commented on framework/13-inch/common/intel.nix:
```diff
@@ -5,6 +5,7 @@
   boot.kernelParams = [
     # Fixes a regression in s2idle, making it more power efficient than deep sleep
+    # Does not really make a difference for me in 12th-gen-intel - Linux 6.8 (SebTM - 04/2024) with/without 0.X Watt usage
     "acpi_osi=\"!Windows 2020\""
```
possibly the s2idle regression has been fixed

huantianad replied:
Did you also test deep vs s2idle power consumption with/without this change as well?

SebTM replied:
@RaitoBezarius Updated based on your comment and will do testing after rebuild/reboot again and add it here.

SebTM commented:
Framework seem to have released the 3.08-bios for the 12th (which was beta for long time): https://knowledgebase.frame.work/framework-laptop-bios-and-driver-releases-12th-gen-intel-core-Bkx2kosqq

but still some quirks - so let's see if we get a statement until the weekend: https://community.frame.work/t/12th-gen-intel-core-bios-3-08-beta-release/43244/263

That might also have an impact on needed changes/configuration.

SebTM commented:
> I was previously trying to figure this out in the oneVPL thread, but should we replace intel-media-driver with onevpl-intel-gpu? Or are they both needed? It was suggested to me that only onevpl is needed, but unsure.

Not sure what you are referring to - nor any knowledge so far so might be a better topic for a separate change?

Mic92 commented on framework/13-inch/12th-gen-intel/default.nix:
```diff
@@ -1,52 +1,67 @@
-      fi
-    '';
-  };
+      # Alder Lake CPUs benefit from kernel 5.18 for ThreadDirector
+      # https://www.tomshardware.com/news/intel-thread-director-coming-to-linux-5-18
+      boot.kernelPackages = lib.mkIf (lib.versionOlder pkgs.linux.version "5.18") (lib.mkDefault pkgs.linuxPackages_latest);
```
I don't think there is any NixOS version left with linux 5.18 as the default. So we can drop this.

SebTM replied:
Dropped as suggested 🙏🏻

Mic92 commented on framework/13-inch/12th-gen-intel/default.nix:
```diff
@@ -1,52 +1,67 @@
+    }
+    # https://community.frame.work/t/tracking-hard-freezing-on-fedora-36-with-the-new-12th-gen-system/20675/391
+    (lib.mkIf (lib.versionOlder config.boot.kernelPackages.kernel.version "6.2") {
+      hardware.intelgpu.loadInInitrd = true;
```
Does this not lead to a conflict with the definition in line 10?

SebTM replied:
I don't think it conflicts as it should be enabled for kernel older 6.2 still to be backwards compatible (as I'm unsure and don't have the time to go back that far)

but I changed the value to be conditional to make it better / clearer readable 🙏🏻 lmk what you think ✌🏻

huantianad commented on framework/13-inch/12th-gen-intel/default.nix:
```diff
@@ -1,52 +1,61 @@
-  # Alder Lake CPUs benefit from kernel 5.18 for ThreadDirector
-  # https://www.tomshardware.com/news/intel-thread-director-coming-to-linux-5-18
-  boot.kernelPackages = lib.mkIf (lib.versionOlder pkgs.linux.version "5.18") (lib.mkDefault pkgs.linuxPackages_latest);
+      # Combined with acpi_osi from framework-intel it increases the idle power-usage from 0.X Watt to up to 2.X Watt in my test (SebTM - 04/2024)
```
Perhaps this is me misreading the comment, but is this saying that *removing* the following `boot.kernelParams` option as well as the acpi_osi will increase the idle power usage?

SebTM replied:
It means, if both options this and the acpi_osi from framework-intel on new kernel-versions (e.g. 6.8) are enabled it increases instead of optimizing the power-usage.

See my redone test-results I will post as comment in a few minutes ✌🏻

SebTM commented:
Hey, I redid my tests regarding the `kernelParams` for:
 * "acpi_osi=\"!Windows 2020\""
 * "mem_sleep_default=deep"

with a clean/stable setup:
```
 - system: `"x86_64-linux"`
 - host os: `Linux 6.8.7, NixOS, 24.05 (Uakari), 24.05.20240420.ce010d9`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.18.2`
 - nixpkgs: `/nix/store/bqjyrqnw03qsccqjvl2ihbvrxjxhjwqj-source`
```
(GNOME Shell 45.5, connected via Wifi, power-consumption measured via Aliexpress multi-usb(-a/c) charger only laptop connected - without laptop connected shows 0 Watt)

1. Battery was fully charged
2. I verified before rebuild and after reboot between each test that `kernelParams` are the expected values: "nix eval .\#nixosConfigurations.<hostname>.config.boot.kernelParams"
3. I cleaned my setup with `sudo nix-collect-garbage --delete-old` and `nix-collect-garbage --delete-old` between each test
4. I used suspend in gnome application-drawer to get into my test-state
5. In all tests only alacrity from verifying `kernelParams` was open

---

1. Current PR - "acpi_osi=\"!Windows 2020\"" and "mem_sleep_default=deep" enabled
Max 1.6W - after ~1min between 1.0/1.2/1.4/1.6W

2. Current PR - Only (from the two) "acpi_osi=\"!Windows 2020\"" enabled
Max 1W - after ~1min between 0.4/0.6W

3. Current PR - Only (from the two) "mem_sleep_default=deep" enabled
Max 1.6W - after ~1min between 1.0/1.2/1.4/1.8W

4. Current PR - latest commit (none from the two) enabled
Max 1W - after ~1min between 0.2/0.4/0.6W

5. Latest master (both enabled)
Max 1.8W - after ~1min between 1.0/1.2/1.4/1.6/1.8W

---

In all tests I didn't noticed issues with keyboard-backlight not turning off.

SebTM commented:
Hey, I think I addressed all review-comments and redid the tests with a stable test-setup which is a bit time-costly so I would appreciate if someone could verify the results on their side so that we can get this merged soon :)

huantianad commented:
Sorry, I might be a bit confused, but if using `mem_sleep_default=deep` increases power draw, why do we still have it included in the config?

SebTM commented:
> Sorry, I might be a bit confused, but if using mem_sleep_default=deep increases power draw, why do we still have it included in the config?

In combination with other options and certain kernel-versions it can but to stay backwards-compatible I only change it from 6.8 upwards - not sure how to test/get more data without extreme time-invest

SebTM commented:
> In terms of overall changes, looks good to me, might be worth rebasing and squashing the sleep changes though.

Done

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/monitor-sensor-doesnt-work-hyprland/37879/3
